File: sdks/go/function/openapi_test.go
package function

import (
	"context"
	"testing"

	"github.com/cuihairu/croupier/sdks/go/pkg/croupier"
	"github.com/getkin/kin-openapi/openapi3"
)

// mockClient is a mock implementation of croupier.Client for testing.
type mockClient struct {
	registeredFunctions map[string]croupier.FunctionDescriptor
}

func newMockClient() *mockClient {
	return &mockClient{
		registeredFunctions: make(map[string]croupier.FunctionDescriptor),
	}
}

func (m *mockClient) RegisterFunction(desc croupier.FunctionDescriptor, handler croupier.FunctionHandler) error {
	m.registeredFunctions[desc.ID] = desc
	return nil
}

func (m *mockClient) Connect(ctx context.Context) error {
	return nil
}

func (m *mockClient) Serve(ctx context.Context) error {
	return nil
}

func (m *mockClient) Stop() error {
	return nil
}

func (m *mockClient) Close() error {
	return nil
}

func TestRegistry_RegisterFromOpenAPI(t *testing.T) {
	// Simple OpenAPI spec
	spec := []byte(`{
		"openapi": "3.0.3",
		"info": {
			"title": "Test API",
			"version": "1.0.0"
		},
		"paths": {
			"/players/ban": {
				"post": {
					"operationId": "player.ban",
					"summary": "Ban Player",
					"tags": ["player", "moderation"],
					"description": "Ban a player from the game",
					"x-category": "player",
					"x-risk": "high",
					"x-permission": "player.ban.invoke",
					"requestBody": {
						"required": true,
						"content": {
							"application/json": {
								"schema": {
									"type": "object",
									"properties": {
										"playerId": {"type": "string"},
										"reason": {"type": "string"}
									}
								}
							}
						}
					},
					"responses": {
						"200": {
							"description": "Success",
							"content": {
								"application/json": {
									"schema": {
										"type": "object",
										"properties": {
											"success": {"type": "boolean"}
										}
									}
								}
							}
						}
					}
				}
			}
		}
	}`)

	client := newMockClient()
	registry := NewRegistryWithLogger(client, &NoOpLogger{})

	// Handler function
	handlerFunc := func(operationID string) Handler {
		return func(ctx context.Context, input []byte) ([]byte, error) {
			return []byte(`{"success":true}`), nil
		}
	}

	options := &ImportOptions{
		ContinueOnError:  false,
		DefaultTimeoutMs: 30000,
	}

	err := registry.RegisterFromOpenAPI(spec, options, handlerFunc)
	if err != nil {
		t.Fatalf("RegisterFromOpenAPI failed: %v", err)
	}

	// Verify the function was registered
	metadata, exists := registry.GetMetadata("player.ban")
	if !exists {
		t.Fatal("Function 'player.ban' was not registered")
	}

	if metadata.Name != "Ban Player" {
		t.Errorf("Expected name 'Ban Player', got '%s'", metadata.Name)
	}

	if metadata.Category != "player" {
		t.Errorf("Expected category 'player', got '%s'", metadata.Category)
	}

	if metadata.Risk.Level != RiskHigh {
		t.Errorf("Expected risk level RiskHigh, got %v", metadata.Risk.Level)
	}
}

func TestRegistry_RegisterFromOpenAPIWithHandlers(t *testing.T) {
	spec := []byte(`{
		"openapi": "3.0.3",
		"info": {
			"title": "Test API",
			"version": "1.0.0"
		},
		"paths": {
			"/test": {
				"get": {
					"operationId": "test.function",
					"summary": "Test Function",
					"responses": {
						"200": {
							"description": "Success"
						}
					}
				}
			}
		}
	}`)

	client := newMockClient()
	registry := NewRegistryWithLogger(client, &NoOpLogger{})

	handlers := map[string]Handler{
		"test.function": func(ctx context.Context, input []byte) ([]byte, error) {
			return []byte(`{}`), nil
		},
	}

	err := registry.RegisterFromOpenAPIWithHandlers(spec, nil, handlers)
	if err != nil {
		t.Fatalf("RegisterFromOpenAPIWithHandlers failed: %v", err)
	}

	_, exists := registry.GetMetadata("test.function")
	if !exists {
		t.Fatal("Function 'test.function' was not registered")
	}
}

func TestRegistry_RegisterFromOpenAPI_WithPrefix(t *testing.T) {
	spec := []byte(`{
		"openapi": "3.0.3",
		"info": {
			"title": "Test API",
			"version": "1.0.0"
		},
		"paths": {
			"/test": {
				"get": {
					"operationId": "test.function",
					"tags": ["tag1", "tag2"],
					"x-category": "mycategory",
					"responses": {
						"200": {
							"description": "Success"
						}
					}
				}
			}
		}
	}`)

	client := newMockClient()
	registry := NewRegistryWithLogger(client, &NoOpLogger{})

	options := &ImportOptions{
		CategoryPrefix: "prefix",
		TagPrefix:      "pre:",
	}

	handlerFunc := func(operationID string) Handler {
		return func(ctx context.Context, input []byte) ([]byte, error) {
			return []byte(`{}`), nil
		}
	}

	err := registry.RegisterFromOpenAPI(spec, options, handlerFunc)
	if err != nil {
		t.Fatalf("RegisterFromOpenAPI failed: %v", err)
	}

	metadata, _ := registry.GetMetadata("test.function")

	if metadata.Category != "prefix.mycategory" {
		t.Errorf("Expected category 'prefix.mycategory', got '%s'", metadata.Category)
	}

	if len(metadata.Tags) != 2 {
		t.Errorf("Expected 2 tags, got %d", len(metadata.Tags))
	}

	if len(metadata.Tags) > 0 && metadata.Tags[0] != "pre:tag1" {
		t.Errorf("Expected tag 'pre:tag1', got '%s'", metadata.Tags[0])
	}
}

func TestParseRiskLevel(t *testing.T) {
	tests := []struct {
		input    string
		expected RiskLevel
	}{
		{"low", RiskLow},
		{"safe", RiskLow},
		{"medium", RiskMedium},
		{"moderate", RiskMedium},
		{"high", RiskHigh},
		{"danger", RiskDanger},
		{"critical", RiskDanger},
		{"unknown", RiskMedium},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			result := parseRiskLevel(tt.input)
			if result != tt.expected {
				t.Errorf("Expected %v, got %v", tt.expected, result)
			}
		})
	}
}

func TestToTitleCase(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"player_ban", "Player Ban"},
		{"get_player_info", "Get Player Info"},
		{"", ""},
		{"single", "Single"},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			result := toTitleCase(tt.input)
			if result != tt.expected {
				t.Errorf("Expected '%s', got '%s'", tt.expected, result)
			}
		})
	}
}

func TestDeriveOperationID(t *testing.T) {
	op := &openapi3.Operation{OperationID: "my.function"}

	// Test with operationID set
	result := deriveOperationID(op, "/some/path")
	if result != "my.function" {
		t.Errorf("Expected 'my.function', got '%s'", result)
	}

	// Test with path only
	op2 := &openapi3.Operation{}
	result2 := deriveOperationID(op2, "/api/players/{id}")
	if result2 != "api.players.{id}" {
		t.Errorf("Expected 'api.players.{id}', got '%s'", result2)
	}
}

File: sdks/js/src/openapi.test.ts
/**
 * Tests for the Descriptor v2 OpenAPI import helper.
 */

import {
  FunctionDescriptor,
  FunctionHandler,
  registerFromOpenAPI,
  ImportOptions,
} from "./index";
import type { RegistrationTarget } from "./openapi";

class RecordingClient implements RegistrationTarget {
  readonly descriptors = new Map<string, FunctionDescriptor>();
  readonly handlers = new Map<string, FunctionHandler>();

  registerFunction(
    descriptor: FunctionDescriptor,
    handler: FunctionHandler,
  ): void {
    this.descriptors.set(descriptor.id, descriptor);
    this.handlers.set(descriptor.id, handler);
  }
}

const SPEC = {
  openapi: "3.0.3",
  info: { title: "GM API", version: "1.0.0" },
  paths: {
    "/players/{id}/ban": {
      put: {
        operationId: "player_ban",
        summary: "Ban player",
        description: "Bans a player account",
        tags: ["gm"],
        "x-resource": "player",
        "x-operation": "ban",
        "x-capability": "action",
        "x-execution": "sync",
        "x-permission": "player.ban",
        "x-risk": "high",
        "x-approval": { required: true, policyKey: "player-ban-v2" },
        requestBody: {
          content: {
            "application/json": {
              schema: {
                type: "object",
                required: ["playerId"],
                properties: {
                  playerId: { type: "string", description: "Player ID" },
                },
              },
            },
          },
        },
        responses: {
          200: {
            content: {
              "application/json": {
                schema: { type: "object", properties: { ok: { type: "boolean" } } },
              },
            },
          },
        },
      },
    },
    "/players/export": {
      post: {
        operationId: "player_export",
        description: "Exports players",
        "x-resource": "player",
        "x-capability": "report",
        "x-execution": "task",
        "x-risk": "low",
        "x-approval": { required: false },
        responses: {
          200: { content: { "application/json": { schema: { type: "object" } } } },
        },
      },
    },
    "/players": {
      get: {
        tags: ["query"],
        "x-capability": "collection_query",
        "x-risk": "warning",
        responses: {
          200: { content: { "application/json": { schema: { type: "array" } } } },
        },
      },
    },
  },
};

function makeHandlers(...ids: string[]): Map<string, FunctionHandler> {
  return new Map(ids.map((id) => [id, ((): FunctionHandler => async () => "{}")()]));
}

describe("registerFromOpenAPI (Descriptor v2)", () => {
  it("maps v2 extensions onto lowerCamelCase descriptor fields", () => {
    const client = new RecordingClient();
    registerFromOpenAPI(
      client,
      SPEC,
      undefined,
      undefined,
      makeHandlers("player_ban", "player_export", "players"),
    );

    const descriptor = client.descriptors.get("player_ban")!;
    expect(Object.keys(descriptor)).toEqual(
      expect.arrayContaining([
        "inputSchema",
        "outputSchema",
        "approvalRequired",
        "approvalPolicyKey",
        "capability",
        "execution",
      ]),
    );
    expect(descriptor.resource).toBe("player");
    expect(descriptor.operation).toBe("ban");
    expect(descriptor.capability).toBe("action");
    expect(descriptor.execution).toBe("sync");
    expect(descriptor.permission).toBe("player.ban");
    expect(descriptor.risk).toBe("high");
    expect(descriptor.approvalRequired).toBe(true);
    expect(descriptor.approvalPolicyKey).toBe("player-ban-v2");
    expect(descriptor.inputSchema).toEqual({
      type: "object",
      required: ["playerId"],
      properties: { playerId: { type: "string", description: "Player ID" } },
    });
    expect(descriptor.outputSchema).toEqual({
      type: "object",
      properties: { ok: { type: "boolean" } },
    });
  });

  it("normalizes legacy risk aliases to the v2 vocabulary", () => {
    const client = new RecordingClient();
    registerFromOpenAPI(
      client,
      SPEC,
      undefined,
      undefined,
      makeHandlers("player_ban", "player_export", "players"),
    );

    expect(client.descriptors.get("player_export")!.risk).toBe("safe");
    expect(client.descriptors.get("player_export")!.approvalRequired).toBe(false);
    expect(client.descriptors.get("player_export")!.approvalPolicyKey).toBeUndefined();
    expect(client.descriptors.get("players")!.risk).toBe("warning");
  });

  it("derives id from the path and title-cases the name when operationId is missing", () => {
    const client = new RecordingClient();
    registerFromOpenAPI(
      client,
      SPEC,
      undefined,
      undefined,
      makeHandlers("player_ban", "player_export", "players"),
    );

    const descriptor = client.descriptors.get("players")!;
    expect(descriptor.id).toBe("players");
    expect(descriptor.name).toBe("Players");
    expect(descriptor.summary).toBe("Players");
    expect(descriptor.capability).toBe("collection_query");
  });

  it("uses summary as name and operationId title-case as fallback", () => {
    const client = new RecordingClient();
    registerFromOpenAPI(
      client,
      SPEC,
      undefined,
      undefined,
      makeHandlers("player_ban", "player_export", "players"),
    );

    expect(client.descriptors.get("player_ban")!.name).toBe("Ban player");
    expect(client.descriptors.get("player_export")!.name).toBe("Player Export");
  });

  it("applies import options including defaultTimeoutMs", () => {
    const client = new RecordingClient();
    const options: ImportOptions = {
      resourcePrefix: "game",
      tagPrefix: "svc-",
      defaultTimeoutMs: 60000,
    };
    registerFromOpenAPI(
      client,
      SPEC,
      options,
      undefined,
      makeHandlers("player_ban", "player_export", "players"),
    );

    const descriptor = client.descriptors.get("player_ban")!;
    expect(descriptor.resource).toBe("game.player");
    expect(descriptor.tags).toEqual(["svc-gm"]);
    expect(descriptor.timeoutMs).toBe(60000);
  });

  it("throws on invalid capability values", () => {
    const client = new RecordingClient();
    expect(() =>
      registerFromOpenAPI(
        client,
        {
          paths: {
            "/players": {
              get: {
                operationId: "player_list",
                "x-capability": "list",
                responses: { 200: { description: "ok" } },
              },
            },
          },
        },
        undefined,
        undefined,
        makeHandlers("player_list"),
      ),
    ).toThrow("convert operation player_list failed: invalid x-capability");
  });

  it("skips invalid operations when continueOnError is set", () => {
    const client = new RecordingClient();
    const registered = registerFromOpenAPI(
      client,
      {
        paths: {
          "/bad": {
            get: {
              operationId: "bad_fn",
              "x-execution": "async",
              responses: { 200: { description: "ok" } },
            },
          },
          "/good": {
            get: {
              operationId: "good_fn",
              "x-execution": "sync",
              responses: { 200: { description: "ok" } },
            },
          },
        },
      },
      { continueOnError: true },
      undefined,
      makeHandlers("bad_fn", "good_fn"),
    );

    expect(registered).toEqual(["good_fn"]);
    expect(client.descriptors.has("good_fn")).toBe(true);
  });

  it("throws on malformed x-approval extensions", () => {
    const client = new RecordingClient();
    expect(() =>
      registerFromOpenAPI(
        client,
        {
          paths: {
            "/players": {
              post: {
                operationId: "player_kick",
                "x-approval": "yes",
                responses: { 200: { description: "ok" } },
              },
            },
          },
        },
        undefined,
        undefined,
        makeHandlers("player_kick"),
      ),
    ).toThrow("x-approval for player_kick must be an object");
  });

  it("registers handlers resolved by function id", () => {
    const client = new RecordingClient();
    registerFromOpenAPI(
      client,
      SPEC,
      undefined,
      undefined,
      makeHandlers("player_ban", "player_export", "players"),
    );

    expect(client.handlers.has("player_ban")).toBe(true);
    expect(client.handlers.has("player_export")).toBe(true);
    expect(client.handlers.has("players")).toBe(true);
  });
});
